# WSI - Ćwiczenie 2

*Autor: Maksymilian Nowak*

### Cel ćwiczenia

Celem ćwiczenia jest zaimplementowanie algorytmu ewolucyjnego:
- z selekcją turniejową ($k=2$)
- krzyżowaniem jednopunktowym
- mutacją gaussowską
- sukcesją generacyjną

Należy również wykorzystać ten algorytm do znalezienia minimum sumy funkcji $f_1+f_2$, gdzie: 
$$f_1(x_1,y_1)=(x_1^2+y_1-11)^2+(x_1+y_1^2-7)^2$$
$$f_2(x_2,y_2)=2*x_2^2+1.05*x_2^4+\frac{x_2^6}{6}+x_2y_2+y_2^2$$
dla dziedziny $D=[-5,5]\times[-5,5]$, oraz zbadać wpływ dystrybucji populacji początkowej, prawdopodobieństwa mutacji oraz prawdopodobieństwa krzyżowania na współrzędne znalezionych minimów funkcji $f_1$.

## Implementacja algorytmu ewolucyjnego

Do implementacji algorytmu przyjąłem następujące założenia:
- Funkcja celu ma następującą postać: $g(x_1, x_2, y_1, y_2)=f_1(x_1,y_1)+f_2(x_2,y_2)$
- Osobnik jest postaci $(x_1,x_2,y_1,y_2)$
- Kryterium stopu w algorytmie będzie maksymalna liczba pokoleń

Na początku zdefiniuję badane funkcje oraz funkcję celu, będącą ich sumą:

In [2]:
def f1(x, y):
    return (x**2 + y - 11)**2 + (x + y**2 - 7)**2

def f2(x, y):
    return 2*x**2 + 1.05*x**4 + (1/6)*x**6 + x*y + y**2

def g(x1, x2, y1, y2):
    return f1(x1, y1) + f2(x2, y2)

Pozostałymi hiperparametrami algorytmu będą:
- prawdopodobieństwo mutacji
- siła mutacji 
- prawdopodobieństwo krzyżowania

Przed zaimplementowaniem algorytmu wyznaczę funkcję oceny - wartością zwracaną przez tę funkcję będzie wynik funkcji celu (im mniejszy, tym lepsza ocena). Dodatkowo kara za wyjście poza dziedzinę funkcji przez osobnika to dodanie 1000 do wyniku działania funkcji oceny.

In [3]:
def rate(genes):
    penalty = 0
    domain_conditions = [-5 <= gene <= 5 for gene in genes]
    if not all(domain_conditions):
        penalty += 1000
    return g(genes[0], genes[1], genes[2], genes[3]) + penalty

Następnie zdefiniuję funkcje odpowiedzialne za reprodukcję, krzyżowanie i mutację, z których będzie się składał algorytm ewoulucyjny.

In [4]:
import random
import numpy as np

# Reprodukcja - selekcja turniejowa (turnieje dwuosobnikowe)
def reproduction(population, rating):
    results = []
    for i, individual in enumerate(population):
        opponents = population.copy()
        opponents.pop(i)
        opponent = random.choice(opponents)
        if rating(individual) < rating(opponent):
            results.append(individual)
        else:
            results.append(opponent)
    return results

# Krzyżowanie jednopunktowe - każda para rodziców produkuje dwójkę dzieci
def crossing(population, crossover_prob):
    results = []
    parents_list = population.copy()
    for i in range(0, len(population) - 1, 2):
        crossover_value = random.uniform(0, 1)
        parent1, parent2 = random.sample(parents_list, 2)
        parents_list.remove(parent1)
        parents_list.remove(parent2)
        if crossover_value < crossover_prob:
            crossing_point = random.randint(1, len(parent1))
            child1, child2 = [
                parent1[:crossing_point] + parent2[crossing_point:],
                parent2[:crossing_point] + parent1[crossing_point:]
            ]
            results.append(child1)
            results.append(child2)
        else:
            results.append(parent1)
            results.append(parent2)
    # Jeśli populacja jest nieparzysta, przenieś rodzica bez pary do wyników
    if len(population) % 2 == 1:
        results.append(parents_list[0])
    return results

# Mutacja gaussowska
def mutation(population, mutation_strength, mutation_prob):
    results = []
    for individual in population:
        mutation_value = random.uniform(0, 1)
        if mutation_value < mutation_prob:
            results.append((individual + mutation_strength * np.random.normal(0, 1, len(individual))).tolist())
        else:
            results.append(individual)
    return results


#### Zaimplementowany algorytm ewolucyjny

In [5]:
def g_minimum(population, rating, crossover_prob, mutation_strength, mutation_prob, max_generations):
    generation = 0
    ratings = [rating(individual) for individual in population]
    best_value = min(ratings)
    best_value_index = ratings.index(best_value)
    best_individual = population[best_value_index]
    while generation < max_generations:
        reproducted = reproduction(population, rating) # Reprodukcja
        crossed = crossing(reproducted, crossover_prob) # Krzyżowanie
        mutated = mutation(crossed, mutation_strength, mutation_prob) # Mutowanie
        # Ocena nowego pokolenia
        new_ratings = [rating(individual) for individual in mutated]
        new_best_value = min(new_ratings)
        new_best_value_index = new_ratings.index(new_best_value)
        new_best_individual = mutated[new_best_value_index]
        if new_best_value < best_value:
            best_value = new_best_value
            best_individual = new_best_individual
        # Sukcesja generacyjna
        population = mutated
        ratings = new_ratings
        generation += 1
    return [best_individual, best_value]

### Szukanie rozwiązania

Przy szukaniu rozwiązania będę wykorzystywał zawsze wykorzystywał populację w postaci 20 osobników o identycznych wartościach genów rosnących od -5 do 5.
Poszukiwania rozpocznę od wartości prawdopodobieństw równych 0,5 oraz siły mutacji równej 0,1.
Maksymalna liczba pokoleń początkowo wyniesie 100.

Ze względu na obecność czynnika losowego przy działaniu algorytmu, dla każdego zestawu hiperparametrów algorytm będzie uruchamiany 50 razy, a następnie zwrócone zostaną:
- uśrednione wartości poszczególnych genów osobników w populacji
- średnia otrzymana wartość funkcji celu
- najlepsza (minimalna) otrzymana wartość tej funkcji
- najgorsza otrzymana wartość tej funkcji
- odchylenie standardowe otrzymanych wyników

In [22]:
x1 = np.linspace(-5, 5, 20)
x2 = np.linspace(-5, 5, 20)
y1 = np.linspace(-5, 5, 20)
y2 = np.linspace(-5, 5, 20)
individuals = list(map(list, zip(x1, x2, y1, y2)))
print(f"Populacja początkowa: {individuals}")

def find_min(population, rate, crossover_prob, mutation_strength, mutation_prob, max_generations, no_of_experiments):
    individual_results = []
    value_results = [] 
    for i in range(no_of_experiments):
        result = g_minimum(population, rate, crossover_prob, mutation_strength, mutation_prob, max_generations)
        individual_results.append(result[0])
        value_results.append(result[1])
    gene_values = list(zip(*individual_results))
    average_individual = [
        sum(gene_values[0]) / len(gene_values[0]),
        sum(gene_values[1]) / len(gene_values[1]),
        sum(gene_values[2]) / len(gene_values[2]),
        sum(gene_values[3]) / len(gene_values[3]),
    ]
    average_value = sum(value_results) / len(value_results)
    print("Uśrednione wartości genów osobników w postaci [x1, x2, y1, y2]:", average_individual)
    print("Średnia wartość funkcji celu:", average_value)
    print("Najlepsza znaleziona wartość", min(value_results))
    print("Najgorsza znaleziona wartość", max(value_results))
    print("Odchylenie standardowe wartości osobników:", np.std(value_results))

Populacja początkowa: [[-5.0, -5.0, -5.0, -5.0], [-4.473684210526316, -4.473684210526316, -4.473684210526316, -4.473684210526316], [-3.947368421052632, -3.947368421052632, -3.947368421052632, -3.947368421052632], [-3.4210526315789473, -3.4210526315789473, -3.4210526315789473, -3.4210526315789473], [-2.8947368421052633, -2.8947368421052633, -2.8947368421052633, -2.8947368421052633], [-2.368421052631579, -2.368421052631579, -2.368421052631579, -2.368421052631579], [-1.8421052631578947, -1.8421052631578947, -1.8421052631578947, -1.8421052631578947], [-1.3157894736842106, -1.3157894736842106, -1.3157894736842106, -1.3157894736842106], [-0.7894736842105265, -0.7894736842105265, -0.7894736842105265, -0.7894736842105265], [-0.2631578947368425, -0.2631578947368425, -0.2631578947368425, -0.2631578947368425], [0.2631578947368416, 0.2631578947368416, 0.2631578947368416, 0.2631578947368416], [0.7894736842105257, 0.7894736842105257, 0.7894736842105257, 0.7894736842105257], [1.3157894736842106, 1.31

In [26]:
find_min(individuals, rate, 0.5, 0.1, 0.5, 100, 50)

Uśrednione wartości genów osobników w postaci [x1, x2, y1, y2]: [2.794383868324374, 0.001594156374692229, 1.8895572170364607, 0.00443016217400295]
Średnia wartość funkcji celu: 0.010658030995110247
Najlepsza znaleziona wartość 0.0006513288803193153
Najgorsza znaleziona wartość 0.0316430984256575
Odchylenie standardowe wartości osobników: 0.0070796481779974415


Otrzymane rozwiązanie znajduje się blisko prawdziwego minimum, którego wartość wynosi 0 (minimalną wartość funkcji celu sprawdziłem za pomocą programu WolframAlpha).
Spróbuję teraz zmienić wartość prawdopodobieństwa krzyżowania z 0,5 na 0,8 i powtórzyć eksperyment:

In [20]:
find_min(individuals, rate, 0.8, 0.1, 0.5, 100, 50)

Uśrednione wartości genów osobników w postaci [x1, x2, y1, y2]: [2.4110362584809826, -0.004648991610248427, 1.9074257195713438, -0.004170288806610022]
Średnia wartość funkcji celu: 0.006658828718358707
Najlepsza znaleziona wartość 0.000668062276788902
Najgorsza znaleziona wartość 0.01875371333841764
Odchylenie standardowe wartości osobników: 0.004150494169005561


Po zmianie margines błędu algorytmu znacząco zmalał. Teraz zbadam, czy zwiększenie liczby pokoleń wpłynie na otrzymywane wyniki - ustawię wartość tego hiperparametru na 200:

In [21]:
find_min(individuals, rate, 0.8, 0.1, 0.5, 200, 50)

Uśrednione wartości genów osobników w postaci [x1, x2, y1, y2]: [2.7785008001772824, 0.0015864559047004414, 1.968175411286864, -0.0036452172579436878]
Średnia wartość funkcji celu: 0.003718408255666829
Najlepsza znaleziona wartość 0.0005263978560194726
Najgorsza znaleziona wartość 0.009230593840580071
Odchylenie standardowe wartości osobników: 0.002329967888543761


Zgodnie z moimi oczekiwaniami, większa liczba pokoleń wpłynęła pozytywnie na otrzymywany wynik. Otrzymane rezultaty uważam za zadowalające, dlatego przejdę teraz do badania wpływu poszczególnych hiperparametrów na algorytm.

##### Tabela z otrzymanymi rezultatami

| Prawdopodobieństwo krzyżowania | Siła mutacji | Prawdopodobieństwo mutacji | Liczba pokoleń | Średni wynik | Najlepszy wynik | Najgorszy wynik | Odchylenie standardowe |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 0.5 | 0.1 | 0.5 | 100 | 0.011 | 0.001 | 0.032 | 0.007 |
| 0.8 | 0.1 | 0.5 | 100 | 0.007 | 0.001 | 0.019 | 0.004 |
| 0.8 | 0.1 | 0.5 | 200 | 0.004 | 0.001 | 0.009 | 0.002 |

## Badanie wpływu hiperparametrów na algorytm

Podczas przeprowadzania eksperymentów po każdym etapie działania algorytmu ewolucyjnego - reprodukcji, krzyżowania i mutacji - będę generował wykres przedstawiający dystrybucję osobników na planszy - każda funkcja będzie posiadała swój osobny wykres. W tym celu zdefiniuję nową funkcję, która wciąż będzie realizowała algorytm ewolucyjny, ale dodatkowo rysowała wykresy: